### PR TITLE
[Sema] Shepherd @manavgabhawala's changes to circular protocol diagnostics

### PR DESCRIPTION
**Pull request by CodaFi**
#### What's in this pull request?

#2123's type checker changes were causing an infinite loop, but the rest of the changes related to circular protocol diagnostics are still valuable to have.  This will pull them in and leave the type checker alone for now.

h/t @rintaro 

---

Before merging this pull request to apple/swift repository:
- [ ] Test pull request on Swift continuous integration.

<details>
  <summary>

Triggering Swift CI</summary>

The swift-ci is triggered by writing a comment on this PR addressed to the GitHub user @swift-ci. Different tests will run depending on the specific comment that you use. The currently available comments are:

**Smoke Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please smoke test |
| All supported platforms | @swift-ci Please smoke test and merge |
| OS X platform | @swift-ci Please smoke test OS X platform |
| Linux platform | @swift-ci Please smoke test Linux platform |

A smoke test on macOS does the following:
1. Builds the compiler incrementally.
2. Builds the standard library only for macOS. Simulator standard libraries and
   device standard libraries are not built.
3. lldb is not built.
4. The test and validation-test targets are run only for macOS. The optimized
   version of these tests are not run.

A smoke test on Linux does the following:
1. Builds the compiler incrementally.
2. Builds the standard library incrementally.
3. lldb is built incrementally.
4. The swift test and validation-test targets are run. The optimized version of these
   tests are not run.
5. lldb is tested.

**Validation Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please test |
| All supported platforms | @swift-ci Please test and merge |
| OS X platform | @swift-ci Please test OS X platform |
| OS X platform | @swift-ci Please benchmark |
| Linux platform | @swift-ci Please test Linux platform |

**Lint Testing**

| Language | Comment |
| --- | --- |
| Python | @swift-ci Please Python lint |

Note: Only members of the Apple organization can trigger swift-ci.
</details>

### CODE REVIEW

**Comment by CodaFi** [comment]
@swift-ci please test and merge.

**Comment by CodaFi** [comment]
@swift-ci please test.

**Comment by CodaFi** [comment]
@swift-ci please test OS X platform.

File: validation-test/compiler_crashers/00695-swift-availabilityattr-isunavailable.swift [hunk 1]
```diff
@@ -5,7 +5,7 @@
 // See http://swift.org/LICENSE.txt for license information
 // See http://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
 
-// RUN: not %target-swift-frontend %s -parse
+// RUN: not --crash %target-swift-frontend %s -parse
```
Comment (jrose-apple) [inline validation-test/compiler_crashers/00695-swift-availabilityattr-isunavailable.swift]: Hm. Is it worth figuring out why this one regressed?

Comment (CodaFi) [inline validation-test/compiler_crashers/00695-swift-availabilityattr-isunavailable.swift]: It crashes somewhere in member lookup (I will rename the test) this time around - I'll take a look at it later today.

Comment (CodaFi) [inline validation-test/compiler_crashers/00695-swift-availabilityattr-isunavailable.swift]: Nevermind, false alarm.  I must have moved the wrong tests.

**Comment by CodaFi** [comment]
Let's get some sanity here.

@swift-ci please test OS X platform.

**Comment by CodaFi** [comment]
This passed linux tests before I futzed around with the crash tests.

**Pull request merged**